cirq/contrib/quirk/cells/input_rotation_cells.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Iterable, Sequence, Iterator
+
+import numpy as np
+
+import cirq
+from cirq import ops, linalg, value
+from cirq.contrib.quirk.cells.cell import Cell, CellMaker
+
+
+class InputRotationCell(Cell):
+    """Applies an operation that depends on an input gate."""
+
+    def __init__(self, identifier: str,
+                 register: Optional[Sequence['cirq.Qid']],
+                 base_operation: 'cirq.Operation', exponent_sign: int):
+        self.identifier = identifier
+        self.register = None if register is None else tuple(register)
+        self.base_operation = base_operation
+        self.exponent_sign = exponent_sign
+
+    def with_input(self, letter, register):
+        if self.register is None and letter == 'a':
+            if isinstance(register, int):
+                raise ValueError('Dependent operation requires known length '
+                                 'input; classical constant not allowed.')
+            return InputRotationCell(self.identifier, register,
+                                     self.base_operation, self.exponent_sign)
+        return self
+
+    def controlled_by(self, qubit: 'cirq.Qid'):
+        return InputRotationCell(self.identifier, self.register,
+                                 self.base_operation.controlled_by(qubit),
+                                 self.exponent_sign)
+
+    def operations(self) -> 'cirq.OP_TREE':
+        if self.register is None:
+            raise ValueError(f"Missing input 'a'")
+        return QuirkInputRotationOperation(self.identifier, self.register,
+                                           self.base_operation,
+                                           self.exponent_sign)
+
+
+@value.value_equality
+class QuirkInputRotationOperation(ops.Operation):
+    """Operates on target qubits in a way that varies based on an input qureg.
+    """
+
+    def __init__(self, identifier: str, register: Iterable['cirq.Qid'],
+                 base_operation: 'cirq.Operation', exponent_sign: int):
+        self.identifier = identifier
+        self.register = tuple(register)
+        self.base_operation = base_operation
+        self.exponent_sign = exponent_sign
```

Comment:
IIUC, `exponent_sign` should either be -1 or +1. If so, then this should raise `ValueError` otherwise.

(And if I'm wrong, then you should rename the parameter.)

Reply:
Added check.